### ISSUE TITLE
Not fully loaded area resulting in broken map

### ISSUE THREAD

**Issue by noxifoxi** [Dadido3/noita-mapcap#1]
While the mod overall worked fine, Noita was a bit slow loading the map fully before the screenshot was made. This results in some faulty chunks on the stitched image.

Is there a way around it (like waiting for the screen portion to fully load or speeding up the map generation/loading on Noitas side)? I was using the the settings on noita-dev.exe given in [README.md](https://github.com/Dadido3/noita-mapcap/blob/master/README.md#how-to-do-a-full-map-capture-with-minimal-trouble).
[My image](https://www.easyzoom.com/image/159643) looks like shit in comparison to [the example one](https://easyzoom.com/image/159431).

**[The broken image.](https://www.easyzoom.com/image/159643)**

My PC specs are pretty beefy (on the CPU side at least):

- **CPU** AMD Ryzen 7 3700X
- **RAM** 32 GB TridentZ 3200 MHz
- **GPU** Nvidia GTX 1060 6GB

I don't think my specs are an issue.

**Comment by Dadido3** [comment]
It's hard to tell why noita loads chunks slowly on your system. No idea if you are using an SSD or if that even matters when noita generates the world.

Unfortunately i haven't found a way to tell if the map is fully loaded, so i have to rely on timing (ingame frames). There are a few ways to to reduce missing chunks:
- Increase `CAPTURE_DELAY` inside of `.../Noita/mods/noita-mapcap/files/capture.lua`.
- Reduce the resolution of the game, as described in the `README.md`.
- In the `F7` debug menu disable or enable `mTrailerMode`. Even though enabling it made things worse here.
- Change the framerate back to `"60"`, or test any other value.

You have to test if one of these or a combination are helping to fix the problem. Once i find a solution that works better, or when API functions are introduced that help to fix those problems, i'll update the mod. For now everyone has to tweak the parameters a bit if there are artifacts.

**Comment by NoxNebula** [comment]
I'm using an Samsung 850 EVO, SATA-SSD, but I was sure this doesn't even matter.
I will try different framerates and/or increasing `CAPTURE_DELAY`, because I was already using 840x840 ingame resolution according to the `README.de`. I'm not sure what `mTrailerMode` does, but I had it disabled.

**Comment by NoxNebula** [comment]
I still don't know what `mTrailerMode` does, but keeping it enabled helped a lot having the entire game window filled, but made the whole process a bit slower. Also increasing `STREAMING_CHUNK_TARGET` to **32** helped a little bit. I lowered the framerate to 300, this might also helped. But having the framerate on 60 wasn't helpful at all and just made the capturing process slower.


I had a weird result when I tried 30 fps with a `CAPTURE_DELAY` of **60** and `mTrailerMode` disabled. The game window was fully filled and generated after moving the view to the desired coordinates. But just in the moment the capture started, some portion of the game window got blacked out, resulting in those weird rectangular holes on the full map.
I'm not sure why but it has something to do with the capturing process.

btw, my taskmanager while capturing a full map:
![tskmng](https://user-images.githubusercontent.com/293639/69719610-24ef4c00-1111-11ea-9d6c-897f27b73cb5.PNG)

**Comment by Dadido3** [comment]
It's strange that Noita behaves so different on your side, your machine is definitely not too slow.

Normally 100% CPU load shouldn't be a problem, atleast here. In that case the game will run slower, but the world tiles get loaded after some ingame frames anyway. Perhaps Noita has some "problem" with that many cores, as in that it creates many worker threads, but the limiting factor is the IO or RAM throughput. In that case it could be that some of the worker threads don't get a tile ready in time, but i have no idea how it really works inside.

As i wrote in discord, you can try to disable some multithreading related settings, as in set `single_threaded_loading="1"` inside of your `config.xml`. There are some more in the in `magic_numbers.xml`:

- `USE_CUSTOM_THREADPOOL`
- `DEBUG_THREADED_WORLD_CREATION`

If you create a map the next time, you can experiment with those.

Just as a sidenote, saving of the screenshots uses a hardcoded amount of 4 worker threads. That shouldn't be a problem on your machine either.

Regarding `STREAMING_CHUNK_TARGET`, i assume it has something to do with the amount of chunks/tiles that are generated/loaded simultaneously.
And `mTrailerMode` seems to cause chunks to be recalculated/refreshed every (few) frame(s), but similar to `STREAMING_CHUNK_TARGET` it's pure speculation and not tested thoroughly. Only the Noita developers can really .

**Comment by NoxNebula** [comment]
I tried a few options in order to render the Nightmare mode, as you can [see here](https://www.easyzoom.com/image/162965).
`single_threaded_loading="1"` in `config.xml` made the whole process slower, increased my CPU load on all cores about 20% and still produced faulty images. So I dropped that idea after a few tries.
For the capture this time I switched back to 600 fps and set `CAPTURE_DELAY = 40`.
It almost worked out but there are still a few spots wrong (besides the broken map generation of Noita).

Maybe I have to increase the capture delay further, but every frame more waiting will result in a much longer capture session.

The next time I will experiment with `USE_CUSTOM_THREADPOOL` and `DEBUG_THREADED_WORLD_CREATION`.

**Comment by NoxNebula** [comment]
I captured another one of the updated Nightmare mode and got really good results but to the cost of a really long capture process:

- **capture.lua**
```
CAPTURE_DELAY = 50
CAPTURE_BIGJUMP_DELAY = 40
```
- **magic_numbers.xml**
```
VIRTUAL_RESOLUTION_X="840" 
VIRTUAL_RESOLUTION_Y="840" 
STREAMING_CHUNK_TARGET="16" 
```
- **config.xml**
```
backbuffer_height="840" 
backbuffer_width="840" 
framerate="600" 
internal_size_h="840" 
internal_size_w="840" 
window_h="840" 
window_w="840"
```
- **debug menu**
`F5` + `F8` + `F12`
`mTrailerMode` enabled

I got about 5 faulty images, which where easy to delete and to recapture.
The not so exiting part about this is the time it took. I spent **06:09 hours** on the full map capturing process, in my opinion this is way to high.

**tl;dr**
Very good results with these settings but it was very very slow.

**Comment by Dadido3** [comment]
Thanks for sharing your settings, it will help people with similar problems.

Btw, you don't need to delete faulty images if the other images overlapping parts are captured correctly. The stitcher will filter those faulty images automatically.

**Comment by NoxNebula** [comment]
> The stitcher will filter those faulty images automatically.
I know, but in my experience this these parts will be darker and I want a perfect image. _OCD kicks in_

But I think this capturing time is unacceptable and I need to optimize something somewhere, if I'm able to...

**Comment by Dadido3** [comment]
One workaround i could think of is to add a mode to the stitcher to detect and remove faulty images automatically. This way you could recapture the faulty images (But with a higher `CAPTURE_DELAY` setting)

**Comment by Dadido3** [comment]
Add a cleanup mode to the stitcher, see [its README.md](/Dadido3/noita-mapcap/blob/master/bin/stitch/README.md)

This should get rid of most missing chunks:

``` Shell Session
./stitch -cleanup 0.999
```

**Comment by NoxNebula** [comment]
The cleanup mode is actually a life saver, really good job.

I talked a bit with the developers, Petri fixed the assault errors and it doesn't seem to crash anymore. I also asked if its possible to implement an API-function to check if the screen area is done generating. Maybe they'll implement such thing, I'll let you know after they answered.

**Comment by Dadido3** [comment]
Such an API function would be great, it would speed up capturing tremendously and fix all the problems. Thanks for asking the devs.

**Comment by Dadido3** [comment]
All the problems should be gone now, as i use DoesWorldExistAt() now. Also capturing should be much faster than before, as there are no delays anymore.

**Comment by noxifoxi** [comment]
Good job @Dadido3, unluckily the Noita devs didn't replied any of my messages anymore and I got a bit mad about that fact, that I stopped following the development of Noita and the mod API.
